torch/overrides.py

```diff
@@ -1785,6 +1787,37 @@ def _no_torch_function_mode() -> Iterator[None]:
         _set_torch_function_mode(old)
 
 
+class TorchFunctionModeInfo(_ModeInfo):
+    def __init__(self):
+        super().__init__(mode_name="torch_function", mode_class=TorchFunctionMode,
+                         base_mode_class=BaseTorchFunctionMode,
+                         required_fn="__torch_function__")
+
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or\
+            isinstance(mode, self.mode_class) or\
+            (isinstance(mode, type) and not issubclass(mode, self.mode_class))
```

Comment:
What if I changed my mind and said I don't want direct Tensor-like modes anymore. Then it would be possible to share this code exactly between the two variants (with only `mode_class` varying).

Reply:
and then the allowed types error message would be simpler